tools/deny.toml

```diff
@@ -0,0 +1,207 @@
+    # The triple can be any string, but only the target triples built in to
+    # rustc (as of 1.40) can be checked against actual config expressions
+    #{ triple = "x86_64-unknown-linux-musl" },
+    # You can also specify which target_features you promise are enabled for a
+    # particular target. target_features are currently not validated against
+    # the actual valid features supported by the target architecture.
+    #{ triple = "wasm32-unknown-unknown", features = ["atomics"] },
+]
+
+# This section is considered when running `cargo deny check advisories`
+# More documentation for the advisories section can be found here:
+# https://embarkstudios.github.io/cargo-deny/checks/advisories/cfg.html
+[advisories]
+# The path where the advisory database is cloned/fetched into
+db-path = "~/.cargo/advisory-db"
+# The url(s) of the advisory databases to use
+db-urls = ["https://github.com/rustsec/advisory-db"]
+# The lint level for security vulnerabilities
+vulnerability = "deny"
+# The lint level for unmaintained crates
+unmaintained = "deny"
+# The lint level for crates that have been yanked from their source registry
+yanked = "deny"
+# The lint level for crates with security notices. Note that as of
+# 2019-12-17 there are no security notice advisories in
+# https://github.com/rustsec/advisory-db
+notice = "warn"
+# A list of advisory IDs to ignore. Note that ignored advisories will still
+# output a note when they are encountered.
+ignore = [
+    "RUSTSEC-2021-0127", # serde_cbor unmaintained; only used in tests
+]
+# Threshold for security vulnerabilities, any vulnerability with a CVSS score
+# lower than the range specified will be ignored. Note that ignored advisories
+# will still output a note when they are encountered.
+# * None - CVSS Score 0.0
+# * Low - CVSS Score 0.1 - 3.9
+# * Medium - CVSS Score 4.0 - 6.9
+# * High - CVSS Score 7.0 - 8.9
+# * Critical - CVSS Score 9.0 - 10.0
+#severity-threshold =
+
+# This section is considered when running `cargo deny check licenses`
+# More documentation for the licenses section can be found here:
+# https://embarkstudios.github.io/cargo-deny/checks/licenses/cfg.html
+[licenses]
+# The lint level for crates which do not have a detectable license
+unlicensed = "deny"
+# List of explicitly allowed licenses
+# See https://spdx.org/licenses/ for list of possible licenses
+# [possible values: any SPDX 3.11 short identifier (+ optional exception)].
+allow = [
+    "MIT",
+    "Apache-2.0",
+    # "Zlib",
+    #"Apache-2.0 WITH LLVM-exception",
+]
+# Allow 1 or more licenses on a per-crate basis, so that particular licenses
+# aren't accepted for every possible crate as with the normal allow list
+exceptions = [
+    # Each entry is the crate and version constraint, and its specific allow list
+    { name = "bindgen", allow = ["BSD-3-Clause"], version = "*" },
+    { name = "instant", allow = ["BSD-3-Clause"], version = "*" },
+    { name = "libloading", allow = ["ISC"], version = "*" },
+]
```

Comment:
These licenses should probably be allowed instead?

Reply:
One possible stance is that we explicitly allow "permissive" licenses such as [MIT](https://opensource.org/licenses/MIT), [Apache-2](https://opensource.org/licenses/Apache-2.0), [Zlib](https://opensource.org/licenses/Zlib), [BSD-3](https://opensource.org/licenses/BSD-3-Clause), [ISC](https://opensource.org/licenses/ISC) -- in my understanding, they should be compatible with MIT and generally don't restrict godot-rust users.

Strictly speaking, BSD-3-Clause has this paragraph:
> 2. Redistributions **in binary form** must reproduce the above copyright notice, this list of conditions and the following disclaimer in the documentation and/or other materials provided with the distribution.

This might include games. I don't know however where this ends, since bindgen is not part of the end-user application, but used during codegen of godot-rust itself. So maybe that's a possible reason to no not generally whitelist all those licenses...

On the other hand, we also use BSD-3 for `instant`:
```
instant v0.1.10
├── parking_lot v0.11.2
│   ├── gdnative-async v0.9.3
```
which we can't reasonably avoid.

So yes, maybe we might as well just allow it. Not sure what a good approach is here. Practice is again different from theory...

Reply:
> Strictly speaking, BSD-3-Clause has this paragraph:
> 
> > 2. Redistributions **in binary form** must reproduce the above copyright notice, this list of conditions and the following disclaimer in the documentation and/or other materials provided with the distribution.
> 
> This might include games. I don't know however where this ends, since bindgen is not part of the end-user application...

I think there's likely a misunderstanding here. End distributions are always required to include the copyright notices of *all* the transitive dependencies, even MIT ones:

> The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

See the Godot Editor's About page for an instant example.

From the same EmbarkStudio also comes [`cargo-about`](https://github.com/EmbarkStudios/cargo-about), a utility for generating such collections of copyright notices, by the way.

Reply:
Ah, thanks for clarifying, makes sense! I added the licenses to the allowed list.

Also good hint regarding `cargo-about`. Maybe at some point we could mention this in the book, under "Exporting a game" or something.